### PR TITLE
Iox #208 add comparison between string and char

### LINKED ISSUE #208: Additional functionality for `cxx::string`
## Brief feature description

We require additional methods so that `cxx::string` is usable like `std::string`

## Detailed information
 * add method `cxx::string::clear` to clear the string
 
 * add method `cxx::string::operator[]()` to access a specific character of the string

 * assignment of `cxx::string` of different capacity
    ```
    iox::cxx::string<100> a;
    iox::cxx::string<29> b;
    a = b; // this should work

    iox::cxx::string<50> c(b); // this should work
    iox::cxx::string<50> c(a); // this should not work
    ```

  * `operator+=` and `operator+` for easy concatenation
    ```
    iox::cxx::string<100> a("hello");
    iox::cxx::string<100> b("world);

    a += b; // this should work
    iox::cxx::string<100> c = a + "world"; // this should work
    ```

  * `substr`
    ```
    iox::cxx::string<100> a("hello world");
    auto helloString = a.substr(0, 5); // should return iox::cxx::string<100> which contains "hello"
    auto worldString = a.substr(6); // should return iox::cxx::string<100> which contains "world"
    ```

  * `find`, `find_first_of`, `find_last_of` optionally

## Tasks

- [x] `find`, `find_first_of`, `find_last_of`
- [x] `operator+=`, `operator+`
- [x] assignment of `cxx::string` of different capacity
- [x] comparision operator between `cxx::string` and string literal or `char[]`
- [x] `operator[]`
- [x] `cxx::string::clear`
- [x] `insert`
- [ ] `append` + `concatenate` of simple chars
- [x] check if it's possible to have a `constexpr cxx::string`

## Refactoring tasks

- ~~[ ] check if it makes sense to define `using Self = string<Capacity>` in the header and use it within methods~~ - only two occurrences
- ~~[ ] in tests: replace `/// @note` with `// BEGIN tests for method xyz` to enable code folding on some IDEs~~ - seems to work only for Eclipse IDE and KDevelop, therefore postponed

### PR DESCRIPTION
**Pull request by FerdinandSpitzschnueffler**
## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Branch follows the naming format (`iox-#123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit messages are signed (`git commit -s`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/advanced/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/release-notes/iceoryx-unreleased.md

## Notes for Reviewer
The compare method is now a template so that a cxx::string can be compared with another cxx::string (like before), a std::string or char array. Additionally, the comparison operators for std::string and char array were implemented. The comparison operators for char ptr have been removed. They once were added because of the implicit conversion from char ptr to std::string. This is now excluded via 'enable_if'. Furthermore, comparison operators for simple chars are added.

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
    - [x] Each unit test case has a unique UUID
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- Partially closes #208

### CODE REVIEW

**Review by elfenpiff — COMMENTED** [review]
Everything reviewed - except the test. They will follow soon.

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/string.inl [hunk 1]
```diff
@@ -214,61 +214,222 @@ inline bool string<Capacity>::unsafe_assign(const std::string& str) noexcept
...
+{
+    auto result = memcmp(c_str(), &other, 1U);
+    if (result == 0)
     {
-        return 1;
+        if (m_rawstringSize > 1U)
+        {
+            return 1;
+        }
+        else
```
Comment (MatthiasKillat) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/string.inl:355]: The `< 1U` case would be the empty string case and for this we would always correctly compare -1 if `other != \0`. Hence it appears we can omit the case, but there is the following problem:

If `this` is the empty string (size 0) and `other` is  `\0` we would get result 0 and also return 0 which is wrong. I may miss something here but I am fairly certain.

Lexicographically the empty string is smaller than any character, in particular `\0` (which is not equal to the empty string). I think the logic is incorrect in this case and we need to properly handle the empty string. The easiest would be an initial size check (there may be more efficient ways in terms of control flow).

**Review by MatthiasKillat — COMMENTED** [review]
Reviewed all but the tests. I will review the tests separately.

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/string.inl [hunk 2]
```diff
@@ -214,61 +214,222 @@ inline bool string<Capacity>::unsafe_assign(const std::string& str) noexcept
...
+template <typename T>
+inline
+    typename std::enable_if<(is_cxx_string<T>::value || is_char_array<T>::value || std::is_same<T, std::string>::value),
+                            int64_t>::type
+    string<Capacity>::compare(const T& other) const noexcept
 {
-    uint64_t otherSize = other.size();
-    if (m_rawstringSize < otherSize)
+    uint64_t otherSize = internal::GetSize<T>::call(other);
+    auto result = memcmp(c_str(), internal::GetData<T>::call(other), std::min(m_rawstringSize, otherSize));
```
[the remark this answers is not in the document]
Comment (MatthiasKillat) [reply iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/string.inl:224]: One could argue it does not matter here if we do not have early `\0` characters in our string. I am not sure we allow those. If we do, `memcmp` is required as `strncmp` stops at the first `\0` character (or if the n characters where compared).

So they behave differently but it is only relevant if we can have a string with `\0` characters, otherwise they should behave the same and theoretically `memcmp` is better than as it will not check for `\0`.

TLDR: if my assumption regarding `\0` chars is correct, `memcpy` should be preferred for performance.

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp [hunk 3]
```diff
@@ -301,93 +301,128 @@ class string
...
     ///
     /// @param [in] other is the string to compare with self
     ///
     /// @return an integer < 0 if the first character that does not match has a lower value in self than in other, 0 if
     /// the contents of both strings are equal, an integer > 0 if the first character that does not match has a greater
     /// value in self than in other
-    template <uint64_t N>
-    int64_t compare(const string<N>& other) const noexcept;
+    template <typename T>
+    typename std::enable_if<(is_cxx_string<T>::value || is_char_array<T>::value || std::is_same<T, std::string>::value),
```
[the remark this answers is not in the document]
Comment (MatthiasKillat) [reply iceoryx_hoofs/include/iceoryx_hoofs/cxx/string.hpp:312]: Yes, that is what I meant by generating a specific type trait to bundle this.

File: iceoryx_hoofs/test/moduletests/test_cxx_string.cpp [hunk 4]
```diff
@@ -1089,153 +1090,634 @@ TYPED_TEST(stringTyped_test, CompareOperatorGreaterEqResultFalseWithDifferentCap
...
+}
+
+TYPED_TEST(stringTyped_test, CompareOperatorsWithDifferentCharWithDifferentSize)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "c9f2e472-3fbf-4e08-a0ee-6bb725e7c0d7");
+    using MyString = typename TestFixture::stringType;
+    constexpr auto STRINGCAP = MyString::capacity();
+
+    std::string temp1(STRINGCAP + 5U, 'M');
+    string<STRINGCAP + 5U> sut1;
```
Comment (elfenpiff) [inline iceoryx_hoofs/test/moduletests/test_cxx_string.cpp:1641]: Could we rename the sut into better understandable names. Since we compare them always with `testChar` I think the name `sutSmallerTestChar`, `sutEqualTestChar` and `sutGreaterTestChar` would help in understanding the single conditions you verify below.

And maybe we can create 3 `EXPECT_THAT` groups for smaller, equal and greater

File: iceoryx_hoofs/test/moduletests/test_cxx_string.cpp [hunk 5]
```diff
@@ -1089,153 +1090,634 @@ TYPED_TEST(stringTyped_test, CompareOperatorGreaterEqResultFalseWithDifferentCap
...
+TYPED_TEST(stringTyped_test, CompareOperatorsWithDifferentChar)
 {
-    ::testing::Test::RecordProperty("TEST_ID", "99a40704-9d0b-49f9-bd5c-4c87b9f2ba00");
+    ::testing::Test::RecordProperty("TEST_ID", "c818b150-b926-4f3c-8405-6327303f12f6");
     using MyString = typename TestFixture::stringType;
     constexpr auto STRINGCAP = MyString::capacity();
-    string<STRINGCAP> testFixedString("M");
-    std::string testSTDString = "L";
-    EXPECT_THAT(testFixedString != testSTDString, Eq(true));
+    string<STRINGCAP> sut1("M");
```
Comment (elfenpiff) [inline iceoryx_hoofs/test/moduletests/test_cxx_string.cpp:1594]: could you rename this into `sutGreaterTestChar` and `sutSmallerTestChar` so that the `EXPECT_` condition can be easier reviewed.

Comment (elfenpiff) [reply iceoryx_hoofs/test/moduletests/test_cxx_string.cpp:1594]: Please do this also for the tests above where it is applicable.

File: iceoryx_hoofs/test/moduletests/test_cxx_string.cpp [hunk 6]
```diff
@@ -16,6 +16,7 @@
 // SPDX-License-Identifier: Apache-2.0
 
 #include "iceoryx_hoofs/cxx/string.hpp"
+#include "iceoryx_hoofs/platform/platform_correction.hpp"
```
[the remark this answers is not in the document]
Comment (FerdinandSpitzschnueffler) [reply iceoryx_hoofs/test/moduletests/test_cxx_string.cpp:19]: Maybe not. This was just a try to satisfy the msvc, but I found another solution. I will remove it and we will see if the CI complains.

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/string.inl [hunk 7]
```diff
@@ -214,61 +214,222 @@ inline bool string<Capacity>::unsafe_assign(const std::string& str) noexcept
...
+inline typename std::enable_if<(is_char_array<T>::value || std::is_same<T, std::string>::value), bool>::type
+operator>=(const T& lhs, const string<Capacity>& rhs) noexcept
+{
+    return (rhs.compare(lhs) <= 0);
+}
+
+template <uint64_t Capacity>
+inline int64_t string<Capacity>::compare(const char& other) const noexcept
+{
+    auto result = memcmp(c_str(), &other, 1U);
```
[the remark this answers is not in the document]
Comment (FerdinandSpitzschnueffler) [reply iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/string.inl:348]: Yes, I deliberately didn't add the check for an empty string but I forgot to implement a test for this case and will add it.

First I thought that `memcmp` is overkill, too. But the code is more readable without writing the comparison of one char and in case of equality add the size check.

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/string.inl (hunk 1 above)
Comment (FerdinandSpitzschnueffler) [reply iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/string.inl:355]: Actually the first character of an empty string is `\0`. It is default constructed this way and it is also set in the clear method. 
For me, it makes sense to return 0 in this case. Furthermore, if you compare an empty std::string with `"\0"` (note that a char comparison is not allowed) it also returns 0.

Comment (MatthiasKillat) [reply iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/string.inl:328]: That may be the case for `std::string` and we can leave it like this. But in the theory of formal languages (https://en.wikipedia.org/wiki/Empty_string#:~:text=In%20formal%20treatments%2C%20the%20empty,%7C%CE%B5%7C%20%3D%200.) this makes no sense and teh STL is at odds with this.

We can leave it as is for STL consistency.

**Review by MatthiasKillat — APPROVED** [review]
Ok, I think the important test cases are covered. I think we have not tested all the different cases for a string to be lexicographically smaller than another though like 1) same length, some common prefix, different rest, 2) different length, common prefix, different rest etc. This gets combinatorially quite exhausting though.

I think this is also a case where we can (in theory) benefit from a generative testing approach (like I try to use for the ServiceDiscovery). The string is even better suited since the reference algorithm is basically in the `std::string`, so we do not even have to implement it ourselves. It is still complex to generate all the required test cases semi-automatically.

Ideally we would write tests for structures with STL  counterpart in a way like: 1) setup the test case, 2) verify that it behaves like the STL

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/string.inl (hunk 1 above)
Comment (FerdinandSpitzschnueffler) [reply iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/string.inl:332]: @MatthiasKillat @elfenpiff I reconsidered this case and decided to differ here from the std::string behavior. If an empty cxx::string is compared with '\0', compare will return -1.

**Review by elfenpiff — APPROVED** [review]

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/string.inl [hunk 8]
```diff
@@ -321,7 +321,11 @@ inline int64_t string<Capacity>::compare(char other) const noexcept
     auto result = memcmp(c_str(), &other, 1U);
     if (result == 0)
     {
-        if (m_rawstringSize > 1U)
+        if (empty())
+        {
+            return -1;
+        }
+        else if (m_rawstringSize > 1U)
```
Comment (MatthiasKillat) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/string.inl:328]: I would recommend avoiding this kind of `else` branching if there is an early return (which should be used as often as possible for easy to follow code flow and optimization).

No need for changes though and the compiler will optimize the jumps/returns itself in this case.

**Review by MatthiasKillat — APPROVED** [review]
